crates/ark/src/lsp/completions/provide.rs

```diff
@@ -21,12 +24,138 @@ pub(crate) fn provide_completions(
 ) -> Result<Vec<CompletionItem>> {
     log::info!("provide_completions()");
 
-    if let Some(completions) = completions_from_unique_sources(context)? {
+    let node_context = gather_completion_context(context);
+    let no_trailing_parens =
+        check_for_function_value(context, &node_context) || check_for_help(&node_context);
+
+    if let Some(completions) = completions_from_unique_sources(context, no_trailing_parens)? {
         return Ok(completions);
     };
 
     // At this point we aren't in a "unique" completion case, so just return a
     // set of reasonable completions based on loaded packages, the open
     // document, the current workspace, and any call related arguments
-    completions_from_composite_sources(context, state)
+    completions_from_composite_sources(context, state, no_trailing_parens)
+}
+
+#[cfg(test)]
+mod tests {
+    use tower_lsp::lsp_types::InsertTextFormat;
+    use tree_sitter::Point;
+
+    use super::*;
+    use crate::lsp::document_context::DocumentContext;
+    use crate::lsp::documents::Document;
+    use crate::lsp::state::WorldState;
+    use crate::r_task;
+
+    fn point_from_cursor(text: &str) -> (String, Point) {
+        let cursor_pos = text.find('@').unwrap();
+        let text = text.replace('@', "");
+        (text, Point::new(0, cursor_pos))
+    }
+
+    fn find_completion(completions: &[CompletionItem], label: &str) -> CompletionItem {
+        completions
+            .iter()
+            .find(|item| item.label == label)
+            .unwrap()
+            .clone()
+    }
+
+    #[test]
```

Comment:
I decided to test this here after all, as opposed to testing individual completion sources. It feels like we just want to know that the completion is correct, regardless of which helper ultimately provided it. It didn't seem to make sense to test the helpers themselves, since there's nothing particularly tricky about them correctly using the `no_trailing_parens` boolean. Again #681 is going to present a big opportunity to revisit testing that relates to "is the `CompletionItem` correct and is it provided by the expected source(s)?"

Reply:
Used the basic structure of a test I found in #369.